Support IC Low Wasm Memory Hook (#4849)
dfinity/motoko: doc/md/canister-maintenance/memory.md

---
sidebar_position: 5
---

# Memory diagnostics

## Low memory hook

The IC allows to implement a low memory hook, which is a warning trigger when main memory is becoming scarce.

For this purpose, a Motoko actor or actor class instance can implement the system function `onLowMemory()`. This system function is scheduled when canister's free main memory space has fallen below the defined threshold `wasm_memory_threshold`, that is is part of the canister settings. In Motoko, `onLowMemory()` implements the `canister_on_low_wasm_memory` hook defined in the IC specification.

Example of using the low memory hook:
```
actor {
    system func onLowMemory() : async () {
        Debug.print("Low memory!");
    }
}
```

The following properties apply to the low memory hook:
* The execution of `onLowMemory` happens with a certain delay, as it is scheduled as a separate asynchronous message that runs after the message in which the threshold was crossed.

> * Once executed, `onLowMemory` is only triggered again when the main memory free space went above the threshold (e.g. by lowering the threshold or shrinking the main memory through canister reinstallation) and when the free space again fell below the threshold.

Reviewer: ```suggestion
* Once executed, `onLowMemory` is only triggered again when the main memory free space exceeds the threshold (e.g. by lowering the threshold or shrinking the main memory through canister reinstallation) and when the free space again falls below the threshold.
```
Not sure I understand this... should that be an or?

Author: I will clarify. The idea is that it first exceeds and then falls below.

Author: I adjusted it. Hopefully it is clearer now.

* Traps or unhandled errors in `lowmemory` are ignored. Traps only revert the changes done in `lowmemory`.

> * Due to its `async` return type, the `lowmemory` function may send further messages and await results.

Reviewer 2: ```suggestion
* Due to its `async*` return type, the `lowmemory` function may send further messages and `await` results.
```

Author: Thank you!

> * Traps or unhandled errors in `onLowMemory` are ignored. They only revert the changes done in `onLowMemory`.

Reviewer: ```suggestion
* Traps or unhandled errors in `onLowMemory` are ignored. Traps only revert the changes done in `onLowMemory`.
```
